src/topotoolbox/stream_object.py

```diff
@@ -0,0 +1,211 @@
+        'm2', or 'km2'. Default is 'pixels'.
+    threshold : int | float | GridObject | np.ndarray, optional
+        The upslope area threshold for flow accumulation. This can be an
+        integer, float, GridObject, or a NumPy array. If more water than in
+        the threshold has accumulated in a cell, it is part of the stream.
+        Default is 0, which will result in the threshold being generated
+        based on this formula: threshold = (avg^2)*0.01
+        where shape = (n,m).
+
+    Raises
+    ------
+    ValueError
+        If the `units` parameter is not 'pixels', 'mapunits', 'm2', or 'km2'.
+    ValueError
+        If the shape of the threshold does not match the flow object shape.
+        """
+        if not isinstance(flow, FlowObject):
+            err = f"{flow} is not a topotoolbox.FlowObject."
+            raise TypeError(err)
+
+        self.cellsize = flow.cellsize
+        self.shape = flow.shape
+
+        # georeference
+        self.bounds = flow.bounds
+        self.transform = flow.transform
+        self.crs = flow.crs
+
+        # Calculate the are of a cell based on the units argument.
+        if units == 'pixels':
+            cell_area = 1.0
+        elif units == 'm2':
+            cell_area = self.cellsize**2
+        elif units == 'km2':
+            cell_area = (self.cellsize*0.001)**2
+        elif units == 'mapunits':
+            if self.crs is not None:
+                if self.crs.is_projected:
+                    # True so cellsize is in meters
+                    cell_area = self.cellsize**2
+                else:
+                    # False so cellsize is in degrees
+                    pass
+        else:
+            err = (f"Invalid unit '{units}' provided. Expected one of "
+                   f"'pixels', 'mapunits', 'm2', 'km2'.")
+            raise ValueError(err) from None
+
+        # Create the appropriate threshold matrix based on the threshold input.
+        if isinstance(threshold, (int, float)):
+            if threshold == 0:
+                avg = (flow.shape[0] + flow.shape[1])//2
+                threshold = np.full(
+                    self.shape, math.floor((avg**2)*0.01), dtype=np.float32)
+            else:
+                threshold = np.full(self.shape, threshold, dtype=np.float32)
+        elif isinstance(threshold, np.ndarray):
+            if threshold.shape != self.shape:
+                err = (f"Threshold array shape {threshold.shape} does not "
+                       f"match FlowObject shape: {self.shape}.")
+                raise ValueError(err) from None
+            threshold = threshold.astype(np.float32, order='F')
+        else:
+            if threshold.shape != self.shape:
+                err = (f"Threshold GridObject shape {threshold.shape} does "
+                       f"not match FlowObject shape: {self.shape}.")
+                raise ValueError(err) from None
+
+            threshold = threshold.z.astype(np.float32, order='F')
```

Comment:
When a user passes an array to `STREAMobj` in MATLAB, it is interpreted as a Boolean array that defines the pixels that are included as part of the stream network. Imagine a user constructing a flow accumulation raster on their own and then thresholding it themselves. In Python, I imagine something like:

```python
acc = flow.flow_accumulation()
W = acc.z >= 10000
StreamObject(flow,W)
```

Maybe we use a different keyword argument for the array `W` in this case, Something like `stream_pixels`?

In the constructor of the `StreamObject`, you use this array instead of checking `acc[i] >= threshold`.

It works like this in case the user has a different definition of the stream network than the flow accumulation threshold. Maybe it comes from a digitized map or another hydrological model.

What you have implements a spatially varying threshold. I think your implementation is correct, but it does not match the MATLAB interface, which I think we should follow in this case. If a user does want to supply a spatially varying threshold, the Boolean array method described above will let them do that by computing the spatially varying threshold array on their own, and supplying `acc >= threshold` as the `W` array above.

Reply:
The way this is implemented right now ignores a potential `threshold` if a `stream_pixels` is provided. Is this intended or should the threshold still be used? I assume not, since like in your example: `W = acc.z >= 10000` accounts for the threshold already.

Reply:
This is correct. If a user supplies a `stream_pixels` array, then we should ignore the threshold. I think how you have it warning the user if they pass both is reasonable.